### PR TITLE
[SPARK-29032][CORE] Simplify Prometheus support by adding PrometheusServlet/Resource

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

Prometheus.io is a CNCF project used widely with K8s.
- https://github.com/prometheus/prometheus

This PR aims to simplify `Prometheus` support by adding `PrometheusServlet` and `PrometheusResource`. The main use cases are `K8s` and `Spark Standalone` cluster environments.

Note that this PR focus Spark-generated metrics. Also, we need to update the document later in a separate PR.

### Why are the changes needed?

There exists a few ways to support `Prometheus`. However, they requires extra configurations and new resources (`port numbers to pull` or `gateways to push`). And, the endpoints are widely spreader over `Master/Slave/Driver/CoarseGrainedExecutorBackend`. This PR aims to export natively Spark's metrics which starts with **metrics_** prefix from the existing (1) method at `Master/Worker/Driver`. For task metrics, instead of using `CoarseGrainedExecutorBackend `, we will reuse the collected information used for the Apache Spark REST API, `api/v1`.
1. Spark JMX Sink + Prometheus JMX Converter
2. Custom Sink + Prometheus Pushgateway

### Does this PR introduce any user-facing change?

Yes. New web interfaces are added along with the existing JSON API.

|              |                         JSON End Point                       |            Prometheus End Point          | 
| ------- | ------------------------------------------- | ---------------------------------- |
| Master | /metrics/master/json/                                     | /metrics/master/prometheus/          |
| Master | /metrics/applications/json/                             | /metrics/applications/prometheus/ |
| Worker | /metrics/json/                                                   | /metrics/prometheus/                      |
| Driver   | /metrics/json/                                                   | /metrics/prometheus/                      |
| Driver   | /api/v1/applications/{id}/(executors\|stages)  | /metrics/detail/prometheus/          |

```
$ bin/spark-shell
...
Spark context Web UI available at http://localhost:4040
...
```

```
$ curl --silent http://localhost:4040/metrics/prometheus/ | head -n5
metrics_local_1568101220707_driver_BlockManager_disk_diskSpaceUsed_MB_Value 0
metrics_local_1568101220707_driver_BlockManager_memory_maxMem_MB_Value 366
metrics_local_1568101220707_driver_BlockManager_memory_maxOffHeapMem_MB_Value 0
metrics_local_1568101220707_driver_BlockManager_memory_maxOnHeapMem_MB_Value 366
metrics_local_1568101220707_driver_BlockManager_memory_memUsed_MB_Value 0
```
### How was this patch tested?

Manually connect the new end-points with `curl`.

Or, run `prometheus --config.file=config.yaml` with the following configuration and see through the Prometheus UI.

**config.yaml**
```yaml
global:                                                                                             
  scrape_interval:     5s                                                                           
  evaluation_interval: 15s                                                                          
  external_labels:                                                                                  
      monitor: 'codelab-monitor'                                                                    
rule_files:                                                                                         
scrape_configs:                                                                                     
  - job_name: 'spark-master'                                                                        
    metrics_path: '/metrics/master/prometheus/'                                                     
    static_configs:                                                                                 
      - targets: ['localhost:8080']                                                                 
  - job_name: 'spark-applications'                                                                  
    metrics_path: '/metrics/applications/prometheus/'                                               
    static_configs:                                                                                 
      - targets: ['localhost:8080']                                                                 
  - job_name: 'spark-worker'                                                                        
    metrics_path: '/metrics/prometheus/'                                                            
    static_configs:                                                                                 
      - targets: ['localhost:8081']                                                                 
  - job_name: 'spark-driver'                                                                        
    metrics_path: '/metrics/prometheus/'                                                            
    static_configs:                                                                                 
      - targets: ['localhost:4040']                                                                 
  - job_name: 'active-executors-and-stages'                                                         
    metrics_path: '/metrics/detail/prometheus/'                                                     
    static_configs:                                                                                 
      - targets: ['localhost:4040']
```

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @vanzin and @srowen . Could you review this please?

**Comment by dongjoon-hyun** [comment]
Also, cc @skonto .

**Comment by dongjoon-hyun** [comment]
Oops. I missed the task metrics. I'll update this PR later.

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala [hunk 1]
```diff
@@ -83,13 +83,15 @@ private[spark] class MetricsSystem private (
...
   private var metricsServlet: Option[MetricsServlet] = None
+  private var prometheusServlet: Option[PrometheusServlet] = None
 
   /**
    * Get any UI handlers used by this metrics system; can only be called after start().
    */
   def getServletHandlers: Array[ServletContextHandler] = {
     require(running, "Can only call getServletHandlers on a running MetricsSystem")
-    metricsServlet.map(_.getHandlers(conf)).getOrElse(Array())
+    metricsServlet.map(_.getHandlers(conf)).getOrElse(Array()) ++
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala]: Can you just write `(metricsServlet ++ prometheusServlet).map...`?

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala [hunk 2]
```diff
@@ -201,6 +203,12 @@ private[spark] class MetricsSystem private (
                 classOf[Properties], classOf[MetricRegistry], classOf[SecurityManager])
               .newInstance(kv._2, registry, securityMgr)
             metricsServlet = Some(servlet)
+          } else if (kv._1 == "prometheusServlet") {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala]: If you want you can make `kv` into `case (key, value)` here for clarity, but not necessary

File: core/src/main/scala/org/apache/spark/metrics/sink/PrometheusServlet.scala [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        sb.append(s"${normalizeKey(k)}Value ${v.getValue}\n")
+      }
+    }
+    registry.getCounters.asScala.foreach { case (k, v) =>
+      sb.append(s"${normalizeKey(k)}Count ${v.getCount}\n")
+    }
+    registry.getHistograms.asScala.foreach { case (k, h) =>
+      val snapshot = h.getSnapshot
+      val prefix = normalizeKey(k)
+      sb.append(prefix + "Count " + h.getCount + "\n")
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/metrics/sink/PrometheusServlet.scala]: I guess you could use interpolation here to be consistent, but it won't matter much.

**Review by srowen — COMMENTED** [review]
What is prometheus BTW? is it worth adding special code for?

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala]: Ur, for now, the first one is `MetricsServlet` and the other is `PrometheusServlet`. `getHandlers` is not inherited from `Sink`.

File: core/src/main/scala/org/apache/spark/metrics/sink/PrometheusServlet.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/metrics/sink/PrometheusServlet.scala]: Thank you. Done.

**Comment by dongjoon-hyun** [comment]
Yes. Prometheus.io is a CNCF project which is used widely with K8s.
- https://prometheus.io

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala (hunk 1 above)
Comment (srowen) [reply core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala]: Oh I see.

**Comment by dongjoon-hyun** [comment]
Although this is not a long PR, but I decided to split this into two PRs because  `PrometheusResource` is a new feature aligned with SPARK-23429 (which is added at 3.0.0). That will make the review easier.

**Comment by dongjoon-hyun** [comment]
Sorry for this change~

**Pull request closed without merging**